### PR TITLE
Corsika output mini master

### PR DESCRIPTION
**Pull request by VictorBarbosaMartins**
The time has come to merge the corsika output module to main. Here you may find:

- the new module `simtools/corsika/corsika_output` which reads a telescope output file from a CORSIKA simulation and allows the production a series of histograms for the Cherenkov photons on the ground. All parameters from the RUN header and EVENT header can also be recovered from the output file;
- Extension and improvements made on general helper functions (e.g. rotation, save file);
- Extensive unit tests for the new module.

This PR is composed of two major PRs, namely #454 #460, already reviewed in details by @orelgueta. 

For the next steps (to be updated on #360):
- Develop the application to use the module (parts of it were already developed for locally testing the module)
- The next tool will have to be able to read the histogram outputs from different corsika output files and compare them with statistical tests.

Please start the review once all tests are passing. Thanks!

Solves #458.

### CODE REVIEW

File: simtools/corsika/corsika_output.py [hunk 1]
```diff
@@ -0,0 +1,1301 @@
...
+
+
+class CorsikaOutput:
+    """CorsikaOutput extracts the Cherenkov photons information from a CORSIKA output file
+    (IACT file) using pyeventio.
+
+    Parameters
+    ----------
+    input_file: str or Path
+        Input file (IACT file) provided by the CORSIKA simulation.
```
Comment (GernotMaier) [inline simtools/corsika/corsika_output.py:34]: Note that in it is 'input file' while in line 28 it is 'output file'. Suggest to remove 'output' in line 28 and talk about 'CORSIKA IACT file' only.

File: simtools/corsika/corsika_output.py [hunk 2]
```diff
@@ -0,0 +1,1301 @@
...
+        self._individual_telescopes = None
+        self._allowed_histograms = {"hist_position", "hist_direction", "hist_time_altitude"}
+        self._allowed_1D_labels = {"wavelength", "time", "altitude"}
+        self._allowed_2D_labels = {"counts", "density", "direction", "time_altitude"}
+        self._header = None
+
+    @property
+    def version(self):
+        """
+        Get the version of the Corsika output file.
```
Comment (GernotMaier) [inline simtools/corsika/corsika_output.py:93]: Here again: could call it `CORSIKA IACT file` or `CORSIKA file`, as this output is the input for this class.

File: simtools/corsika/corsika_output.py [hunk 3]
```diff
@@ -0,0 +1,1301 @@
...
+                    for i_key, key in enumerate(self.all_event_keys[1:]):
+                        self.event_information[key].append(event.header[i_key + 1])
+
+                    self.num_events += 1
+
+                all_event_astropy_units = self._get_header_astropy_units(
+                    self.all_event_keys, all_event_units
+                )
+
+                # Add the unity to dictionary with the parameters and turn it into
```
Comment (GernotMaier) [inline simtools/corsika/corsika_output.py:185]: Do you mean `unity` or should that be `units`?

File: simtools/corsika/corsika_output.py [hunk 4]
```diff
@@ -0,0 +1,1301 @@
...
+        """
+        The configuration of the histograms.
+
+        Returns
+        -------
+        dict:
+            the dictionary with the histogram configuration.
+        """
+        if self._hist_config is None:
+            msg = "No configuration was defined before. The default config is being created now."
```
Comment (GernotMaier) [inline simtools/corsika/corsika_output.py:279]: Would it make sense to say `No histogram configuration...`, as there are other configuration steps in the program and this might be confused.

File: simtools/corsika/corsika_output.py [hunk 5]
```diff
@@ -0,0 +1,1301 @@
...
+        Three histograms are created: hist_position with 3 dimensions (x, y positions and the
+        wavelength), hist_direction with 2 dimensions (directive cosinus in x and y directions),
+        hist_time_altitude with 2 dimensions (time and altitude of emission).
+
+        Four arguments are passed to each dimension in the dictionary:
+
+        "bins": the number of bins,
+        "start": the first element of the histogram,
+        "stop": the last element of the histogram, and
+        "scale" to define the scale of the bins which can be "linear" or "log". If "log",
```
Comment (GernotMaier) [inline simtools/corsika/corsika_output.py:338]: Is log == log_10 or log == ln? I would prefer to see this written explicitly in the comment.

File: simtools/corsika/corsika_output.py [hunk 6]
```diff
@@ -0,0 +1,1301 @@
...
+                },
+            },
+            "hist_time_altitude": {
+                "x axis": {
+                    "bins": 100,
+                    "start": -2000 * u.ns,
+                    "stop": 2000 * u.ns,
+                    "scale": "linear",
+                },
+                "y axis": {"bins": 100, "start": 50 * u.km, "stop": 0 * u.km, "scale": "linear"},
```
Comment (GernotMaier) [inline simtools/corsika/corsika_output.py:397]: Why 50 km and not more? CORSIKA sims start earlier (think at 118 km) and we might miss here Cherenkov photons from the primary particle before interaction (for cosmic rays).

File: simtools/corsika/corsika_output.py [hunk 7]
```diff
@@ -0,0 +1,1301 @@
...
+             cy: direction cosinus in the y direction, i.e., the cosinus of the angle between the
+             incoming direction and the y axis,
+             time: time of arrival of the photon in ns. The clock starts when the particle crosses
+             the top of the atmosphere (CORSIKA-defined) if `self.event_first_interaction_heights`
+             is positive or at first interaction if otherwise.
+             zem: altitude where the photon was generated in cm,
+             photons: number of photons associated to this bunch,
+             wavelength: the wavelength of the photons in nm.
+        azimuth_angle: astropy.Quantity
+            Azimuth angle to rotate the observational plane and obtain it perpendicular to the
```
Comment (GernotMaier) [inline simtools/corsika/corsika_output.py:510]: Don't get this sentence. Is the grammar correct for `..and obtain it perpendicular...`?

File: simtools/corsika/corsika_output.py [hunk 8]
```diff
@@ -0,0 +1,1301 @@
...
+             cx: direction cosinus in the x direction, i.e., the cosinus of the angle between the
+             incoming direction and the x axis,
+             cy: direction cosinus in the y direction, i.e., the cosinus of the angle between the
+             incoming direction and the y axis,
+             time: time of arrival of the photon in ns. The clock starts when the particle crosses
+             the top of the atmosphere (CORSIKA-defined) if `self.event_first_interaction_heights`
+             is positive or at first interaction if otherwise.
+             zem: altitude where the photon was generated in cm,
+             photons: number of photons associated to this bunch,
+             wavelength: the wavelength of the photons in nm.
```
Comment (GernotMaier) [inline simtools/corsika/corsika_output.py:508]: Do you actually get a photon wavelength of the bunch size is != 1?

File: simtools/corsika/corsika_output.py [hunk 9]
```diff
@@ -0,0 +1,1301 @@
...
+        azimuth_angle: astropy.Quantity
+            Azimuth angle to rotate the observational plane and obtain it perpendicular to the
+            incoming event. It can be passed in radians or degrees.
+            If not given, no rotation is performed.
+        zenith_angle: astropy.Quantity
+            Zenith angle to rotate the observational plane and obtain it perpendicular to the
```
Comment (GernotMaier) [inline simtools/corsika/corsika_output.py:514]: Same as comment on azimuth.

File: simtools/corsika/corsika_output.py [hunk 10]
```diff
@@ -0,0 +1,1301 @@
...
+        Extract the information of the Cherenkov photons from a CORSIKA output IACT file, create
+         and fill the histograms
+
+        Parameters
+        ----------
+        telescope_indices: int or list of int
+            The indices of the specific telescopes to be inspected.
+        individual_telescopes: bool
+            if False, the histograms are supposed to be filled for all telescopes. Default is False.
+            if True, one histogram is set for each telescope sepparately.
```
Comment (GernotMaier) [inline simtools/corsika/corsika_output.py:569]: Type - 'separately'

File: simtools/corsika/corsika_output.py [hunk 11]
```diff
@@ -0,0 +1,1301 @@
...
+            The counts of the wavelength histogram.
+        np.array
+            The edges of the wavelength histogram in nanometers.
+
+        """
+        return self._get_hist_1D_projection("wavelength")
+
+    def get_photon_time_distr(self):
+        """
+        Get the distribution of the emitted time of the Cherenkov photons. The clock starts when the
```
Comment (GernotMaier) [inline simtools/corsika/corsika_output.py:912]: Should that be `time of emission`? Otherwise it sounds like C photons emit times.

File: simtools/corsika/corsika_output.py [hunk 12]
```diff
@@ -0,0 +1,1301 @@
...
+        -------
+        numpy.ndarray
+            The counts of the histogram.
+        numpy.array
+            The edges of the time histograms in ns.
+
+        """
+        return self._get_hist_1D_projection("time")
+
+    def get_photon_altitude_distr(self):
```
Comment (GernotMaier) [inline simtools/corsika/corsika_output.py:926]: Was 'distribution' to long for the function name? It looks cut off.

File: simtools/corsika/corsika_output.py
Comment (GernotMaier) [inline simtools/corsika/corsika_output.py:1]: This is the second longest module we have. I think in future we should try to keep modules smaller (<1000 lines), makes it hard to keep the overview (don't change here anything).

File: simtools/util/general.py [hunk 13]
```diff
@@ -770,3 +776,93 @@ def get_log_excerpt(log_file, n_last_lines=30):
...
+    ----------
+    dictionary: dict
+        Dictionary to be saved into a file.
+    file_name: str or Path
+        Name of file to be saved with path.
+    """
+    logger = logging.getLogger(__name__)
+    file_name = Path(file_name).with_suffix(".yml")
+    logger.info(f"Exporting histogram configuration to {file_name}")
+    with open(file_name, "w") as file:
```
Comment (GernotMaier) [inline simtools/util/general.py:867]: No try and except when writing a file?

File: simtools/visualization/visualize.py
Comment (GernotMaier) [inline simtools/visualization/visualize.py:1]: Need to rethink if this module shouldn't be split up. If we add any plotting function here, it will get super long and hard to maintain (it is already a very long module).

File: simtools/util/general.py [hunk 14]
```diff
@@ -770,3 +776,93 @@ def get_log_excerpt(log_file, n_last_lines=30):
         f"{collect_final_lines(log_file, n_last_lines)}\n\n"
         "====================================================================\n"
     )
+
+
+def convert_2D_to_radial_distr(hist2d, xaxis, yaxis, bins=50, max_dist=1000):
```
Comment (GernotMaier) [inline simtools/util/general.py:781]: Bit surprised to see this in `general.py`. It seems to be me a very specific function, and not needed in most of the code. Need to discuss what to do.

File: tests/resources/tel_output_10GeV-2-gamma-20deg-CTAO-South.dat
Comment (GernotMaier) [inline tests/resources/tel_output_10GeV-2-gamma-20deg-CTAO-South.dat:1]: What kind of format is this file? Can see it from the suffix (I usually expect that a .dat file is ascii).

**Review by GernotMaier — CHANGES_REQUESTED** [review]
Mostly minor changes, but two discussion points.

File: simtools/visualization/visualize.py
Comment (orelgueta) [reply simtools/visualization/visualize.py:1]: I made the same comment that I think the detailed plotting should be in `corsika_output` or a new `corsike_output_visualise` module, while `visualize.py` should be just for the basic kernel plotting (without any labels/titles/etc.)

File: simtools/corsika/corsika_output.py (hunk 6 above)
Comment (VictorBarbosaMartins) [reply simtools/corsika/corsika_output.py:397]: For the example file I used 50 km was fine but you are right, the default value should deal with all the cases. I changed it to 120 km now.

File: simtools/corsika/corsika_output.py (hunk 7 above)
Comment (VictorBarbosaMartins) [reply simtools/corsika/corsika_output.py:510]: I improved it now and changed the name of the variables to match the ones from the rotate function.
I am actually not sure what would be best, to rotate the observational plane by default and align it to the incoming direction of particles or not to rotate by default.

File: simtools/corsika/corsika_output.py (hunk 8 above)
Comment (VictorBarbosaMartins) [reply simtools/corsika/corsika_output.py:512]: I am not sure I get the question. 
`photons` gives the number of photons in the bunch and `wavelength` represents the wavelength of all the photons in the bunch.

the specific corsika IACT file that I use in the tests was configured for bunch size 1, so to be honest I am not sure what would happen if this would be >1, I would have to test a different file.

File: simtools/corsika/corsika_output.py (hunk 12 above)
Comment (VictorBarbosaMartins) [reply simtools/corsika/corsika_output.py:933]: not specifically this one but I used 'distr' in all functions and some of them would be quite long with 'distribution'

File: simtools/util/general.py (hunk 13 above)
Comment (VictorBarbosaMartins) [reply simtools/util/general.py:867]: I think I removed it at some point (or really forgot). Isn't sufficient that yaml raises the error if the file does not exist?
If you prefer I can also implement it here

File: simtools/visualization/visualize.py
Comment (VictorBarbosaMartins) [reply simtools/visualization/visualize.py:1]: I could move everything to a new `simtools/visualization/visualize_corsika_output.py` file. Would that be good?

File: simtools/util/general.py (hunk 14 above)
Comment (VictorBarbosaMartins) [reply simtools/util/general.py:781]: This is a generic mathematical conversion (from hist2d to hist1d), I think it might be useful at some point for other 2d distributions

File: tests/resources/tel_output_10GeV-2-gamma-20deg-CTAO-South.dat
Comment (VictorBarbosaMartins) [reply tests/resources/tel_output_10GeV-2-gamma-20deg-CTAO-South.dat:1]: it is Konrad's format for the CORSIKA IACT file (binary file). I had before no extension and Orel preferred me to add one. Do you prefer no extension, other extension, e.g., '.bin' or leave it like this?

File: simtools/corsika/corsika_output.py (hunk 12 above)
Comment (GernotMaier) [reply simtools/corsika/corsika_output.py:933]: OK - we are usually have pretty long names. Good!

File: simtools/util/general.py (hunk 13 above)
Comment (GernotMaier) [reply simtools/util/general.py:867]: Only if the yaml error message is understandable. In general, python packages have very user-unfriendly error messages and are hard to parse. This is why I would prefer a clearer error message from us (but maybe here it is ok, not sure).

File: simtools/visualization/visualize.py
Comment (GernotMaier) [reply simtools/visualization/visualize.py:1]: Not now, better to open an issue and first complete this pull request as it is.

File: tests/resources/tel_output_10GeV-2-gamma-20deg-CTAO-South.dat
Comment (GernotMaier) [reply tests/resources/tel_output_10GeV-2-gamma-20deg-CTAO-South.dat:1]: Would `corsikaio` a better suffix? It would indicate the file format.

File: simtools/util/general.py (hunk 13 above)
Comment (VictorBarbosaMartins) [reply simtools/util/general.py]: Ok, I implemented a quick check and the test for it

File: simtools/corsika/corsika_output.py (hunk 8 above)
Comment (VictorBarbosaMartins) [reply simtools/corsika/corsika_output.py:512]: I opened the issue #472 to address this comment later and make sure we can handle different input configurations than the one used for tests here

**Review by GernotMaier — APPROVED** [review]
All good, thank you!
